wgpu-hal/src/gles/egl.rs

```diff
@@ -683,7 +688,7 @@ enum WindowKind {
 
 #[derive(Clone, Debug)]
 struct WindowSystemInterface {
-    display_owner: Option<Arc<DisplayOwner>>,
+    display_owner: Option<Arc<Mutex<DisplayOwner>>>,
```

Comment:
Why is this a mutex all of a sudden?

This could use an in-code explanation.

Reply:
> Why is this a mutex all of a sudden?

`cargo clippy` gives me the following warning:

<details>

```rust
warning: usage of an `Arc` that is not `Send` or `Sync`
   --> wgpu-hal/src/gles/egl.rs:805:32
    |
805 |                 (display, Some(Arc::new(library)), WindowKind::Wayland)
    |                                ^^^^^^^^^^^^^^^^^
    |
    = note: the trait `Sync` is not implemented for `DisplayOwner`
    = note: required for `Arc<DisplayOwner>` to implement `Send` and `Sync`
    = help: consider using an `Rc` instead or wrapping the inner type with a `Mutex`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arc_with_non_send_sync
    = note: `#[warn(clippy::arc_with_non_send_sync)]` on by default

warning: usage of an `Arc` that is not `Send` or `Sync`
   --> wgpu-hal/src/gles/egl.rs:816:32
    |
816 |                 (display, Some(Arc::new(display_owner)), WindowKind::X11)
    |                                ^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: the trait `Sync` is not implemented for `DisplayOwner`
    = note: required for `Arc<DisplayOwner>` to implement `Send` and `Sync`
    = help: consider using an `Rc` instead or wrapping the inner type with a `Mutex`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arc_with_non_send_sync

warning: usage of an `Arc` that is not `Send` or `Sync`
   --> wgpu-hal/src/gles/egl.rs:833:32
    |
833 |                 (display, Some(Arc::new(display_owner)), WindowKind::AngleX11)
    |                                ^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: the trait `Sync` is not implemented for `DisplayOwner`
    = note: required for `Arc<DisplayOwner>` to implement `Send` and `Sync`
    = help: consider using an `Rc` instead or wrapping the inner type with a `Mutex`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arc_with_non_send_sync
```

</details>

Should I remove it?

> This could use an in-code explanation.

Aye, I can add it.

Reply:
There certainly shouldn't be a `Mutex` that isn't actually being used for interior mutability. Either

* the `Arc` can simply be downgraded to an `Rc`,
* this is a situation clippy doesn't understand and the lint should be `allow`, or
* there should be a (justified!) `unsafe impl Sync for DisplayOwner` to go along with the `unsafe impl Send`.

Since this is supposed to be a lint fix PR, not add new capabilities, the third option is certainly wrong.

(I'm not familiar with this part of the `wgpu` code, so I don't know which is right; I just saw a situation that could use further explanation.)

Reply:
> the Arc can simply be downgraded to an Rc,

Hm... I assumed that there's an `Arc` in order to be able to let multiple threads use the wgpu-api.

> there should be a (justified!) unsafe impl Sync for DisplayOwner to go along with the unsafe impl Send.

`Sync` shouldn't be possible, if I'm right because two threads could call the `get_ptr` function on `DisplayOwner::X11` and then they can write/read/modify this pointer (through `unsafe` though) without "notifying" the other thread(s), but maybe I'm  wrong here.

Reply:
> I assumed that there's an `Arc` in order to be able to let multiple threads use the wgpu-api.

`Arc` doesn't *add* any thread-safety. The only thing it does is add multiple-ownership, like `Rc`. The difference is that using `Rc` loses `Send + Sync` if you already had either, but `Arc` does not. But (in the existing state of the code) you don't have `Send` *or* `Sync`, so there is no benefit to `Arc`. That's what Clippy's trying to tell you: “this `Arc` does you no good”.

Reply:
> > I assumed that there's an `Arc` in order to be able to let multiple threads use the wgpu-api.
> 
> `Arc` doesn't _add_ any thread-safety. The only thing it does is add multiple-ownership, like `Rc`. The difference is that using `Rc` loses `Send + Sync` if you already had either, but using `Arc` does not. But (in the existing state of the code) you don't have `Send` _or_ `Sync`, so there is no benefit to `Arc`. That's what Clippy's trying to tell you: “this `Arc` does you no good”.

oh ok, that sounds reasonable. I replaced `Arc` with `Rc` now.